lib/tree.rb

```diff
@@ -16,47 +16,141 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def add(key, value=nil)
+    new_node = TreeNode.new(key, value)
+    if @root == nil 
+      @root = new_node 
+    else
+      add_helper(@root, new_node)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current == nil
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+
+    return current
+  end
+
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def find(key)
```

Comment:
👍 The time/space complexity is O(log n) if the tree is balanced and O(n) if it is unbalanced.